File: frontend/src/types/LineItem.ts
import { TServiceItem } from './ServiceItem';

export type TLineItem = {
  id: string;
  serviceItemId?: string;
  serviceItem?: TServiceItem;
  name: string;
  description?: string;
  unitPrice: number;
  cost: number;
  isOptional: boolean;
  isTaxable: boolean;
  quantity: number;
  total: number;
  jobId?: string;
  invoiceId?: string;
  quoteId?: string;
  requestId?: string;
  createdAt: string;
  updatedAt: string;
};

export type TGetLineItem = {
  id: string;
  name: string;
  description?: string;
  unitPrice: number;
  quantity: number;
  isOptional: boolean;
  total: number;
};

export type TModalLineItem = {
  serviceItemId: string;
  serviceItem: TServiceItem;
  quantity: number;
};

export type TAddLineItem = {
  serviceItemId?: string;
  name: string;
  description?: string;
  unitPrice: number;
  cost?: number;
  isOptional: boolean;
  isTaxable?: boolean;
  quantity: number;
  jobId?: string;
  invoiceId?: string;
  quoteId?: string;
  requestId?: string;
};

export type TUpdateLineItem = {
  id?: string;
  serviceItemId?: string;
  name: string;
  description?: string;
  unitPrice: number;
  isTaxable?: boolean;
  quantity: number;
};

File: backend/Models/LineItem.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using backend.Models.QuoteModels;
namespace backend.Models;

public class LineItem
{
    [Key]
    public Guid Id { get; set; }
    public Guid? ServiceItemId { get; set; }
    public ServiceItem? ServiceItem { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    [Range(0, double.MaxValue)]
    public decimal UnitPrice { get; set; }
    [Range(0, double.MaxValue)]
    public decimal Cost { get; set; }
    public bool IsOptional { get; set; }
    public bool IsTaxable { get; set; }
    [Range(1, int.MaxValue)]
    public int Quantity { get; set; } = 1;
    [NotMapped]
    public decimal Total => UnitPrice * Quantity;

    public Guid? JobId { get; set; }
    [JsonIgnore]
    public Job? Job { get; set; }
    public Guid? InvoiceId { get; set; }
    [JsonIgnore]
    public Invoice? Invoice { get; set; }
    public Guid? QuoteId { get; set; }
    [JsonIgnore]
    public Quote? Quote { get; set; }
    public Guid? LeadId { get; set; }
    [JsonIgnore]
    public Lead? Lead { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}